quickwit-search/src/export/root.rs

```diff
@@ -0,0 +1,194 @@
+use quickwit_proto::SearchRequest;
+
+use crate::client_pool::Job;
+use crate::error::parse_grpc_error;
+use crate::list_relevant_splits;
+use crate::root::job_for_splits;
+use crate::root::NodeSearchError;
+use crate::ClientPool;
+use crate::SearchClientPool;
+use crate::SearchError;
+
+/// Perform a distributed export.
+pub async fn root_export(
+    export_request: &ExportRequest,
+    metastore: &dyn Metastore,
+    client_pool: &Arc<SearchClientPool>,
+) -> Result<Vec<u8>, SearchError> {
+    let start_instant = tokio::time::Instant::now();
+    // TODO: building a search request should not be necessary for listing splits.
+    // This needs some refactoring: relevant splits, metadata_map, jobs...
+    let search_request = SearchRequest::from(export_request.clone());
+    let split_metadata_list = list_relevant_splits(&search_request, metastore).await?;
+    let split_metadata_map: HashMap<String, SplitMetadata> = split_metadata_list
+        .into_iter()
+        .map(|split_metadata| (split_metadata.split_id.clone(), split_metadata))
+        .collect();
+    let leaf_search_jobs: Vec<Job> =
+        job_for_splits(&split_metadata_map.keys().collect(), &split_metadata_map);
+    let assigned_leaf_search_jobs = client_pool
+        .assign_jobs(leaf_search_jobs, &HashSet::default())
+        .await?;
+
+    debug!(assigned_leaf_search_jobs=?assigned_leaf_search_jobs, "Assigned leaf search jobs.");
+
+    let mut handles = Vec::new();
+    for (mut search_client, jobs) in assigned_leaf_search_jobs {
+        let split_ids: Vec<String> = jobs.iter().map(|job| job.split.clone()).collect();
+        let leaf_export_request = LeafExportRequest {
+            export_request: Some(export_request.clone()),
+            split_ids: split_ids.clone(),
+        };
+        debug!(leaf_export_request=?leaf_export_request, grpc_addr=?search_client.grpc_addr(), "Leaf node export.");
+        let handle = tokio::spawn(async move {
+            let mut stream = search_client
+                .leaf_export(leaf_export_request)
+                .await
+                .map_err(|search_error| NodeSearchError {
+                    search_error,
+                    split_ids: split_ids.clone(),
+                })?;
+
+            let mut leaf_bytes = Vec::new();
+            while let Some(leaf_result) = stream.next().await {
+                let leaf_data = leaf_result.map_err(|status| NodeSearchError {
+                    search_error: parse_grpc_error(&status),
+                    split_ids: split_ids.clone(),
+                })?;
+                leaf_bytes.extend(leaf_data.data);
+            }
+            Result::<Vec<u8>, NodeSearchError>::Ok(leaf_bytes)
+        });
+        handles.push(handle);
+    }
+    let export_results = futures::future::try_join_all(handles).await?;
+    let mut errors = Vec::new();
+    // TODO: refactor this part...
+    let mut buffer = Vec::new();
+    for response in export_results {
```

Comment:
It might be faster to traverse a first time the results, collect the total number of bytes that we should allocate and allocate it first. Then traverse a second time to write the `responses`

Reply:
Currently just returning the `Vec<Vec<u8>>` , and reserving capacity. we could also use partition on `export_result`.